cpp/src/arrow/compute/kernels/set_lookup_internal.h

```diff
@@ -0,0 +1,37 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include "arrow/compute/kernel.h"
+
+namespace arrow::compute::internal {
+
+/// Base class for `is_in` and `index_in` kernel states.
+struct SetLookupStateBase : public KernelState {
+  /// Enables non-templated access to the value set type.
+  std::shared_ptr<DataType> value_set_type;
+  /// Enables fast simplification for `is_in` expressions.
+  ///
+  /// Note that this value not initialized by the kernel and used exclusively
+  /// by `SimplifyWithGuarantee`.
+  ///
+  /// \invariant sorted and contains no null or duplicate values
```

Comment:
I may ask a stupid question but should we limit the type here? Something like:
1. nested type
2. float (with NaN) 
Should not in here?

Reply:
That sounds reasonable. I've changed it so that we only simplify for integer, temporal, and base binary types.